modules/pipeline/providers/leaderworker/leader.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderworker
+
+import (
+	"context"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+
+	"github.com/erda-project/erda/modules/pipeline/providers/leaderworker/worker"
+)
+
+func (p *provider) leaderFramework(ctx context.Context) {
+	p.Log.Infof("start leader framework")
```

Comment:
too complex func, can you separate to some clear meaning small funcs?

Reply:
yes, needed.

Reply:
adjusted to `listenWorkerChange`.